keras/objectives.py

```diff
@@ -6,29 +6,50 @@
 
 epsilon = 1.0e-15
 
-def mean_squared_error(y_true, y_pred):
-    return T.sqr(y_pred - y_true).mean()
+def mean_squared_error(y_true, y_pred, weight=None):
+    if weight is not None:
+        return T.sqr(weight*(y_pred - y_true)).mean()
+    else:
+        return T.sqr(y_pred - y_true).mean()
 
-def mean_absolute_error(y_true, y_pred):
-    return T.abs_(y_pred - y_true).mean()
+def mean_absolute_error(y_true, y_pred, weight=None):
+    if weight is not None:
+        return T.abs_(weight*(y_pred - y_true)).mean()
+    else:
+        return T.abs_(y_pred - y_true).mean()
 
-def squared_hinge(y_true, y_pred):
-    return T.sqr(T.maximum(1. - y_true * y_pred, 0.)).mean()
+def squared_hinge(y_true, y_pred, weight=None):
+    if weight is not None:
+        return T.sqr(T.maximum(1. - weight*(y_true * y_pred), 0.)).mean()
+    else:
+        return T.sqr(T.maximum(1. - y_true * y_pred, 0.)).mean()
 
-def hinge(y_true, y_pred):
-    return T.maximum(1. - y_true * y_pred, 0.).mean()
+def hinge(y_true, y_pred, weight=None):
+    if weight is not None:
+        return T.maximum(1. - weight*(y_true * y_pred), 0.).mean()
+    else:
+        return T.maximum(1. - y_true * y_pred, 0.).mean()
 
-def categorical_crossentropy(y_true, y_pred):
+def categorical_crossentropy(y_true, y_pred, weight=None):
     '''Expects a binary class matrix instead of a vector of scalar classes
     '''
     y_pred = T.clip(y_pred, epsilon, 1.0 - epsilon)
     # scale preds so that the class probas of each sample sum to 1
     y_pred /= y_pred.sum(axis=1, keepdims=True) 
-    return T.nnet.categorical_crossentropy(y_pred, y_true).mean()
+    cce = T.nnet.categorical_crossentropy(y_pred, y_true)
+    if weight is not None:
+        # return avg. of scaled cat. crossentropy
+        return (weight*cce).mean()
+    else:
+        return cce.mean()
 
-def binary_crossentropy(y_true, y_pred):
+def binary_crossentropy(y_true, y_pred, weight=None):
     y_pred = T.clip(y_pred, epsilon, 1.0 - epsilon)
-    return T.nnet.binary_crossentropy(y_pred, y_true).mean()
+    bce = T.nnet.binary_crossentropy(y_pred, y_true).mean()
```

Comment:
Taking the mean makes this quantity a scalar, I believe. It should be a vector (different value for each sample).